lib/trino-filesystem-azure/src/main/java/io/trino/filesystem/azure/AzureMultipartOutputStream.java

```diff
@@ -0,0 +1,248 @@
+        this.overwrite = overwrite;
+        this.memoryContext = memoryContext.newLocalMemoryContext(AzureOutputStream.class.getSimpleName());
+        this.blockClient = blobClient.getBlockBlobClient();
+        this.uploadExecutor = listeningDecorator(requireNonNull(uploadExecutor, "uploadExecutor is null"));
+    }
+
+    @Override
+    public void write(int b)
+            throws IOException
+    {
+        ensureOpen();
+        ensureCapacity(1);
+        buffer[bufferSize] = (byte) b;
+        bufferSize++;
+        flushBuffer(false);
+    }
+
+    @Override
+    public void write(byte[] bytes, int offset, int length)
+            throws IOException
+    {
+        ensureOpen();
+
+        while (length > 0) {
+            ensureCapacity(length);
+
+            int copied = min(buffer.length - bufferSize, length);
+            arraycopy(bytes, offset, buffer, bufferSize, copied);
+            bufferSize += copied;
+
+            flushBuffer(false);
+
+            offset += copied;
+            length -= copied;
+        }
+    }
+
+    @Override
+    public void flush()
+            throws IOException
+    {
+        ensureOpen();
+        flushBuffer(false);
+    }
+
+    private void ensureOpen()
+            throws IOException
+    {
+        if (closed) {
+            throw new IOException("Output stream closed: " + location);
+        }
+    }
+
+    private void ensureCapacity(int extra)
+    {
+        int capacity = min(writeBlockSizeBytes, bufferSize + extra);
+        if (buffer.length < capacity) {
+            int target = max(buffer.length, initialBufferSize);
+            if (target < capacity) {
+                target += target / 2; // increase 50%
+                target = clamp(target, capacity, writeBlockSizeBytes);
+            }
+            buffer = Arrays.copyOf(buffer, target);
+            memoryContext.setBytes(buffer.length);
+        }
+    }
+
+    private void flushBuffer(boolean finished)
+            throws IOException
+    {
+        // skip multipart upload if there would only be one staged block
+        if (finished && !multiStagingStarted) {
+            BlobOutputStream blobOutputStream = blockClient
+                    .getBlobOutputStream(overwrite);
+
+            blobOutputStream.write(buffer, 0, bufferSize);
+            blobOutputStream.close();
+            return;
+        }
+
+        if (bufferSize != writeBlockSizeBytes && (!finished || bufferSize == 0)) {
+            // If the buffer isn't full yet and we are not finished, do not stage the block
+            return;
+        }
+
+        multiStagingStarted = true;
+
+        byte[] data = buffer;
+        int length = bufferSize;
+
+        if (finished) {
+            this.buffer = null;
+        }
+        else {
+            this.buffer = new byte[0];
+            this.initialBufferSize = writeBlockSizeBytes;
+            bufferSize = 0;
+        }
+        memoryContext.setBytes(0);
```

Comment:
Should memoryContext be zeroed before the upload is done ?
Won't we hold on to the memory at least until upload finishes ?

Reply:
No. Buffer field is reset before this line

Reply:
Done